File: Javascript Projects/linear_regression.js
// Sample data
// X: Feature, y: Target
const X = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
const y = [3, 4, 2, 5, 6, 7, 8, 9, 10, 11];

// Function to calculate mean
function mean(values) {
    return values.reduce((sum, value) => sum + value, 0) / values.length;
}

// Function to calculate variance
function variance(values, meanValue) {
    return values.reduce((sum, value) => sum + (value - meanValue) ** 2, 0);
}

// Function to calculate covariance
function covariance(x, meanX, y, meanY) {
    return x.reduce((sum, xi, i) => sum + (xi - meanX) * (y[i] - meanY), 0);
}

// Calculate the mean of X and y
const meanX = mean(X);
const meanY = mean(y);

// Calculate variance and covariance
const varX = variance(X, meanX);
const covarXY = covariance(X, meanX, y, meanY);

// Calculate coefficients
const b1 = covarXY / varX;
const b0 = meanY - b1 * meanX;

// Make predictions
function predict(x) {
    return b0 + b1 * x;
}

// Predictions for the sample data
const yPred = X.map(x => predict(x));

// Calculate Mean Squared Error
const mse = y.reduce((sum, yi, i) => sum + (yi - yPred[i]) ** 2, 0) / y.length;

// Calculate R^2 score
const ssTotal = y.reduce((sum, yi) => sum + (yi - meanY) ** 2, 0);
const ssResidual = y.reduce((sum, yi, i) => sum + (yi - yPred[i]) ** 2, 0);
const r2Score = 1 - (ssResidual / ssTotal);

console.log(`Coefficients: b0 = ${b0}, b1 = ${b1}`);
console.log(`Mean squared error: ${mse}`);
console.log(`Coefficient of determination (R^2): ${r2Score}`);

// Plot outputs using Plotly.js
const trace1 = {
    x: X,
    y: y,
    mode: 'markers',
    type: 'scatter',
    name: 'Data Points'
};

const trace2 = {
    x: X,
    y: yPred,
    mode: 'lines',
    type: 'scatter',
    name: 'Fitted Line'
};

const layout = {
    xaxis: { title: 'X' },
    yaxis: { title: 'y' },
    title: 'Linear Regression Example'
};

Plotly.newPlot('plot', [trace1, trace2], layout);

File: Python Projects/Linear_Regression.py
# Sample data
X = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
y = [3, 4, 2, 5, 6, 7, 8, 9, 10, 11]

# Function to calculate mean
def mean(values):
    return sum(values) / len(values)

# Function to calculate variance
def variance(values, mean_value):
    return sum([(x - mean_value) ** 2 for x in values]) / (len(values) - 1)  # Dividing by N - 1

# Function to calculate sample covariance
def covariance(x, mean_x, y, mean_y):
    covar = 0.0
    for i in range(len(x)):
        covar += (x[i] - mean_x) * (y[i] - mean_y)
    return covar / (len(x) - 1)  # Dividing by N - 1

# Calculate the mean of X and y
mean_x = mean(X)
mean_y = mean(y)

# Calculate variance and covariance
var_x = variance(X, mean_x)
covar_xy = covariance(X, mean_x, y, mean_y)

# Calculate coefficients
b1 = covar_xy / var_x
b0 = mean_y - b1 * mean_x

# Make predictions
def predict(x):
    return b0 + b1 * x

# Predictions for the sample data
y_pred = [predict(x) for x in X]

# Calculate Mean Squared Error
mse = sum([(y[i] - y_pred[i]) ** 2 for i in range(len(y))]) / len(y)

# Calculate R^2 score
ss_total = sum([(yi - mean_y) ** 2 for yi in y])
ss_residual = sum([(y[i] - y_pred[i]) ** 2 for i in range(len(y))])
r2_score = 1 - (ss_residual / ss_total)

print("Coefficients: b0 =", b0, ", b1 =", b1)
print("Mean squared error:", mse)
print("Coefficient of determination (R^2):", r2_score)

# Plot outputs
import matplotlib.pyplot as plt

plt.scatter(X, y, color='black')
plt.plot(X, y_pred, color='blue', linewidth=3)

plt.xlabel('X')
plt.ylabel('y')
plt.title('Linear Regression Example')
plt.show()
